### PR TITLE
Update signature section

### LINKED ISSUE #24: Rework Signatures section
From #23 and side-discussions, it is clear that the signatures section needs to be fundamentally reorganized.

### PR DESCRIPTION
**Pull request by SalusaSecondus**
Update signature section.

Fixes #24

### CODE REVIEW

**Comment by rozbb** [comment]
> Just because a signature is valid for a message and public key doesn't mean that the holder of the private key knows what the message was.

I'm conflicted about whether I think this is a reasonable takeaway. On the one hand: blind signatures do exist. On the other hand, what is a Bitcoin user saying when they sign a transaction, if not "this transaction's contents were observed and approved by me, the secret key holder"?

**Comment by dennisjackson** [comment]
> > Just because a signature is valid for a message and public key doesn't mean that the holder of the private key knows what the message was.
> 
> I'm conflicted about whether I think this is a reasonable takeaway. On the one hand: blind signatures do exist. On the other hand, what is a Bitcoin user saying when they sign a transaction, if not "this transaction's contents were observed and approved by me, the secret key holder"?

The problem here isn't with blind signatures, many commonly used signature schemes allow an attacker to craft a signature valid for a particular message, without knowing the message. See the first table in 'Seems Legit' for examples. 

Imagine a designer builds a pairing protocol where two parties have a shared secret and need to learn each others public keys. They have the parties sign a nonce and the shared secret. This breaks because the attacker doesn't need to know the shared secret to produce a valid signature over it.

File: index.md [hunk 1]
```diff
@@ -144,14 +142,28 @@ If this isn't sufficient for your design, please seek out experts to talk to.
...
+    (This, along with other properties of low-order points, was used to break [Scuttlebutt](https://en.wikipedia.org/wiki/Secure_Scuttlebutt) in section 7.1 of [this paper by Cremers and Jackson](https://eprint.iacr.org/2019/526) from 2020.)
+  * More trivially, signatures are generally over message digests, so if the hash function is broken (such as [MD5](https://eprint.iacr.org/2013/170) or [SHA-1](https://shattered.io/)) then a single signature is also valid for all messages with colliding hashes. [This was used to attack Windows Update in 2012](https://arstechnica.com/information-technology/2012/06/flame-crypto-breakthrough/).
+* Just because a signature is valid for a message and public key doesn't mean that the holder of the private key knows what the message was.
+  * Consider both the universal signature and hash collisions from an earlier point where the signatures are (by definition) valid for an unbounded collection of messages.
+  * There is also the trivial case that the signer might only know the hash and not the real message.
```
Comment (dennisjackson) [inline index.md:165]: Could add: "Some signature schemes (e.g. RSA-PKCSv1.5) allow an attacker to extract the message hash from a signature".

File: index.md [hunk 2]
```diff
@@ -144,14 +142,28 @@ If this isn't sufficient for your design, please seek out experts to talk to.
...
+[Digital Signatures](https://en.wikipedia.org/wiki/Digital_signature) are generally safe to use, but many people assume they have properties that they do not. At their core all they mean is that *without the private key, an attacker cannot find a signature for the associated public key over an arbitrary value for which they don't already know a signature*. (a.k.a. [existential unforgeability])
+The best overview of these additional (unpromised) security properties is in  [Seems Legit: Automated Analysis of Subtle Attacks on Protocols that Use Signatures](https://eprint.iacr.org/2019/779)
+and I strongly recommend that everyone at least skim that paper.
+(There is lots of formal verification in there which you can probably skip, but the introduction, defined properties, and case studies are critical.)
+For an easier read, though focused entired on ECDSA, you should look at [How Not to Use ECDSA](https://yondon.blog/2019/01/01/how-not-to-use-ecdsa/).
+
+That said, here are lots of "gotchas" for digital signatures:
```
Comment (dennisjackson) [inline index.md:151]: Could add: "An attacker may be able to extract information from a signature, such as the the signer's public key or the hash of the signed message."

File: index.md (hunk 1 above)
Comment (SalusaSecondus) [reply index.md:165]: I already have this:
> * Signatures do not (necessarily) hide the message they sign. They'll commonly reveal the hash (or more) of the data being signed.

File: index.md (hunk 2 above)
Comment (SalusaSecondus) [reply index.md:151]: I already have this:
> * Signatures do not (necessarily) hide the message they sign. They'll commonly reveal the hash (or more) of the data being signed.
> * Signatures do not hide the public key used to verify the message.
>   * "Public Key Recovery" is a standardized part of ECDSA. (See [SEC2](https://secg.org/sec1-v2.pdf), Section 4.1.6-4.1.7.)
>   * This is slightly more challenging with RSA but is still doable as [described here](https://crypto.stackexchange.com/a/26190) and [implemented here](https://gist.github.com/divergentdave/40ac9c7224b382166c905e76595bcf73).

Comment (dennisjackson) [reply index.md:151]: Ah sorry, I only looked at the diff, not the full file.

File: index.md (hunk 1 above)
Comment (dennisjackson) [reply index.md:165]: Ah sorry, I only looked at the diff.

File: index.md [hunk 3]
```diff
@@ -160,10 +160,12 @@ That said, here are lots of "gotchas" for digital signatures:
...
-* Just because a signature is valid for a message and public key doesn't mean that the holder of the private key knows what the message was.
-  * Consider both the universal signature and hash collisions from an earlier point where the signatures are (by definition) valid for an unbounded collection of messages.
-  * There is also the trivial case that the signer might only know the hash and not the real message.
-  * This can also be an intentional feature (rather than a bug) as in the case of [Blind Signatures](https://eprint.iacr.org/2019/526)
+* Just because a signature is valid for a message and public key doesn't mean that the holder of the private key necessarily knows what the message was.
+  * This can be unintentional if the underlying hash function is broken (see earlier point)
+  * Otherwise this generally requires intentional behavior by the signer:
+    * Consider the maliciously created universal signature from earlier.
+    * There is the trivial case that the signer might only know the hash and not the real message.
+    * This can also be an intentional feature (rather than a bug) as in the case of [Blind Signatures](https://eprint.iacr.org/2019/526)
```
Comment (rozbb) [inline index.md:168]: Awkward: intentional behavior...intentional feature

**Pull request merged**